plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSig/Stage0.hs

```diff
@@ -0,0 +1,247 @@
+        in valuePreserved'
+
+    ||]
+
+-- | @valuePaid pm ptx@ is true if the pending transaction @ptx@ pays
+--   the amount specified in @pm@ to the public key address specified in @pm@
+valuePaid :: Q (TExp (Payment -> PendingTx -> Bool))
+valuePaid = [||
+
+        let valuePaid' :: Payment -> PendingTx -> Bool
+            valuePaid' (Payment vl pk _) ptx = $$(Value.eq) vl ($$(Validation.valuePaidTo) ptx pk)
+        in valuePaid'
+    ||]
+
+-- | Equality of 'Payment' values
+paymentEq :: Q (TExp (Payment -> Payment -> Bool))
+paymentEq = [||
+
+        let paymentEq' :: Payment -> Payment -> Bool
+            paymentEq' (Payment vl pk sl) (Payment vl' pk' sl') = 
+                $$(P.and) ($$(P.and) ($$(Value.eq) vl vl') ($$(Validation.eqPubKey) pk pk')) ($$(Slot.eq) sl sl')
+        in paymentEq'
+
+    ||]
+
+-- | Equality of 'State' values
+stateEq :: Q (TExp (State -> State -> Bool))
+stateEq = [||
+        let stateEq' :: State -> State -> Bool
+            stateEq' (InitialState v) (InitialState v') =
+                $$(Value.eq) v v'
+            stateEq' (CollectingSignatures vl pmt pks) (CollectingSignatures vl' pmt' pks') =
+                $$(P.and) ($$(P.and) ($$(Value.eq) vl vl') ($$paymentEq pmt pmt')) ($$pkListEq pks pks')
+            stateEq' _ _ = False
+
+        in stateEq'
+    ||]
+
+-- | @step params state input@ computes the next state given current state
+--   @state@ and the input.
+step :: Q (TExp (Params -> State -> Input -> State))
+step = [||
+
+    let step' :: Params -> State -> Input -> State
+        step' p s i = 
+            case (s, i) of
+                (InitialState vl, ProposePayment pmt) -> 
+                    CollectingSignatures vl pmt []
+                (CollectingSignatures vl pmt pks, AddSignature pk) ->
+                    CollectingSignatures vl pmt (pk:pks)
+                (CollectingSignatures vl pmt _, Cancel) ->
+                    InitialState vl
+                (CollectingSignatures vl pmt@(Payment vp _ _) pks, Pay) ->
+                    let vl' = $$(Value.minus) vl vp in
+                    InitialState vl'
+                _ -> $$(P.error) ($$(P.traceH) "invalid transition" ())
+    in step'
+    ||]
+
+-- | @stepWithChecks params ptx state input@ computes the next state given 
+--   current state @state@ and the input. It checks whether the pending 
+--   transaction @ptx@ pays the expected amounts to script and public key 
+--   addresses. Fails with 'P.error' if an invalid transition is attempted.
+stepWithChecks :: Q (TExp (Params -> PendingTx -> State -> Input -> State))
+stepWithChecks = [||
+
+    let and_ :: Bool -> Bool -> Bool
+        and_ = $$(P.and)
+
+        not_ :: Bool -> Bool
+        not_ = $$(P.not)
+        
+        step' :: Params -> PendingTx -> State -> Input -> State
+        step' p ptx s i = 
+            let newState = $$step p s i in
+            case (s, i) of
+                (InitialState vl, ProposePayment pmt) -> 
```

Comment:
Since you've already separated state transitions, you can also separate validation into a function that returns either `ValidationSuccess` or a `ValidationError` and then define `step'` as (pseudocode)

```haskell
step' p ptx s i = case $$validate p ptx s i of
    ValidationSuccess -> $$step p s i
    ValidationError err -> $$(P.error) ($$(P.traceH) err ())
```

But it's not important at all.